lib/backend/options.go

```diff
@@ -0,0 +1,22 @@
+// Copyright (c) 2016-2019 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package backend
+
+// ListOptions defines a the options which can be specified
+// when listing names. It is used to enable pagination in list requests.
+type ListOptions struct {
+	Paginated bool
+	MaxKeys int64
+	ContinuationToken string
+}
```

Comment:
Can we use variadic functional options instead, to avoid having to pass an empty struct on every List callsite? Then we could also use non-zero defaults.

E.g.

```
type ListOptions struct {
  Paginated bool
  MaxKeys int64
  ContinuationToken
}

func DefaultListOptions() *ListOptions {
  return &ListOptions{
    ???
  }
}

type ListOption func(*ListOptions)

func ListWithPagination() ListOption {
  return func(o *ListOptions) { o.Paginated = true }
}

func ListWithMaxKeys(n int64) ListOption {
  return func(o *ListOptions) { o.MaxKeys = n }
}

// ... etc ...
```

Then a List function looks like:

```
func (c *client) List(prefix string, opts ...backend.ListOption) (*backend.ListResult, error) {
  options := backend.DefaultListOptions()
  for _, o := range opts {
    o(options)
  }
  ...
}
```

And a callsite might look like:

```
res, err := client.List("foo", backend.ListWithPagination())
...
```

Reply:
+1

Reply:
I can do that, it does potentially make paginated callsites kinda long (or multi-line) though if you want to set all parameters.

```
res, err := client.List("foo",
    backend.ListWithPagination(),
    backend.ListWithMaxKeys(),
    backend.ListWithContinuationToken()
)

```